tdrs-backend/docker-compose.override.yml

```diff
@@ -0,0 +1,7 @@
+# Local version of Docker compose, using environment variables passed in by env file
```

Comment:
This is what I am seeing:	
1.  Run `docker-compose -f docker-compose.yml build`
2. Run `docker-compose -f docker-compose.yml up`

This results in errors because the JWT_KEY is missing
1. Run `docker-compose build`
2. Run `docker-compose up`
Everything running as expected again 
Does this differ from what you were trying?

Reply:
Steps to reproduce:
- Set `tdpservice/.env` `JWT_KEY` to `foo`
- Add `print(os.environ['JWT_KEY'])` to `tdpservice/wsgi.py`
- Delete `docker-compose.overrride.yml`
- docker-compose -f docker-compose.yml build
- docker-compose -f docker-compose.yml up

Observe in console log that `foo` still is displayed, despite `docker-compose.overrride.yml` not existing.

Reply:
I did a little more digging on this and its actually because of the `JWT_KEY=${JWT_KEY}` line in the base docker-compose. This line tells docker to pull the JWT_KEY from system environment variables, but docker does its own magic under the hood to find this. First, it will check the host's environment variables (anything set on your actual machine by using export, etc). Then it will look for the presence of any file explicitly named `.env` and look for the matching variable declaration in there. This is a separate functionality from the `env_file` directive.

However, if the `.env` file is removed or renamed (and there is no variable defined on the system) the docker build will proceed without raising errors for the missing file but will show that the variables are missing:
```
❯ docker-compose up -d                                                                                                                                                                                          ─╯
WARNING: The JWT_KEY variable is not set. Defaulting to a blank string.
WARNING: The JWT_CERT_TEST variable is not set. Defaulting to a blank string.
```

The distinction with using the `env_file` directive is as follows:
* docker will load *all* environment variables contained in `.env` without having to list each one in docker-compose
* docker will raise an error if the `.env` file is missing
* any filename can be used - it doesn't  have to be explicitly `.env`

It would be functionally equivalent to have one docker-compose file that doesn't use the `env_file` directive but instead defines each variable in docker-compose using interpolation. In that setup the `.env` would only be used if present and we would intentionally not have it present in CI or production. The drawback to that approach is that it is a little less obvious that the `.env` file is being used at all locally and developers could run into issues if they don't define the variable in the compose file. By using the `env_file` directive we cut off a step to add new variables, because they just need to be added to `.env` and then they can be immediately leveraged after restarting the container. It is important to note that for true secrets we will still need to define those in the base docker-compose, so that they can be pulled in from CircleCI Project Settings in the CI environment.

These blog posts have some excellent details if you're interested in a more in depth explanation:
https://vsupalov.com/override-docker-compose-dot-env/
https://vsupalov.com/docker-arg-env-variable-guide/

Reply:
Oh wow, fascinating! So basically the `tdrs-backend/.env` was always being read in by Docker Compose. Thank you for looking closely at this @hursey013 and @jtwillis92!

Just as we didn't need `docker-compose.override.yml`, it seems like we don't really need `docker-compose.local.yml`, either! Since as you mentioned above @jtwillis92, just using `.env` without any extra config is functionally equivalent, with a few small differences. 

@jtwillis92 Which approach do you think is simpler and better from a dev experience perspective: using `env_file` to call `.env`, or just using the default Docker Compose behavior of reading `.env`? I could go either way.

Reply:
I personally prefer using the `env_file` directive because developers can add new environment variables quickly without having to update docker-compose and the `.env` file but its mostly a matter of preference. We can achieve the same functionality either way.

An addition benefit of using the `env_file` directive is that we can easily distinguish *true secrets* from configurable environment variables. Secrets will still need to be defined in the docker-compose file using the dollar notation so we can easily see which values are considered secrets and expected to be passed in from CI context and then all the other environment variables can just go straight into .env and not need to be passed in through docker-compose.

Essentially any variable that looks like this in docker-compose: `- JWT_KEY=${JWT_KEY}` would be a secret and we could use that as a list of variables that must be defined in CircleCI.

Reply:
We can sum it up to the pros and cons of the two choices:

### A single docker-compose using dollar notation
#### Pros
* Simpler docker-compose up command
#### Cons
* Not immediately obvious where variables are coming from
* No errors presented to developers on startup if `.env` is not present
* Each variable that is available must be listed in docker-compose using dollar notation and also defined in `.env`

### A separate docker-compose for local dev that uses `env_file`
#### Pros
* Simpler to add new environment variables
* Expectation of `.env` file being present is clear to devs, and an error will be raised if missing
* Secrets still need to  be defined, so it can be assumed all dollar notation in compose file means a variable is secret
#### Cons
* Longer docker startup command

Reply:
Thanks for listing out the pros and cons, appreciate the thoroughness here! There are more subtleties in the differences between the two approaches than I realized. 

Going with the more explicit `env_file` approach seems like a good developer experience to me! 👍